File: src/ikj/main/ioke/lang/Arity.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang;

import ioke.lang.exceptions.ControlFlow;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Arity extends IokeData {

    private enum Taking { Nothing, Everything };

    private static final IokeObject getArity(IokeObject self, Taking thing) {
        IokeObject obj = self.runtime.arity.allocateCopy(null, null);
        obj.mimicsWithoutCheck(self.runtime.arity);
        Arity arity = new Arity(thing);
        obj.setData(arity);
        return obj;
    }

    public static final IokeObject getArity(IokeObject self, DefaultArgumentsDefinition def) throws ControlFlow {
        if( def == null || def.isEmpty() ) {
            return IokeObject.as(takingNothing(self), self.runtime.arity);
        }
        IokeObject obj = self.runtime.arity.allocateCopy(null, null);
        obj.mimicsWithoutCheck(self.runtime.arity);
        Arity arity = new Arity(def);
        obj.setData(arity);
        return obj;
    }

    private DefaultArgumentsDefinition argumentsDefinition;
    private Taking taking;

    public Arity(Taking taking) {
        this.taking = taking;
    }

    public Arity(DefaultArgumentsDefinition argumentsDefinition) {
        if(argumentsDefinition == null || argumentsDefinition.isEmpty()) {
            this.taking = Taking.Nothing;
        } else {
            this.argumentsDefinition = argumentsDefinition;
        }
    }

    public static final Object takingNothing(IokeObject self) throws ControlFlow {
        return self.runtime.arity.getCell(null, null, "taking:nothing");
    }

    public static final Object takingEverything(IokeObject self) throws ControlFlow {
        return self.runtime.arity.getCell(null, null, "taking:everything");
    }

    @Override
    public void init(final IokeObject arity) throws ControlFlow {
        arity.setKind("Arity");
        
        arity.setCell("taking:nothing", getArity(arity, Taking.Nothing));
        arity.setCell("taking:everything", getArity(arity, Taking.Everything));
        
        arity.registerMethod(arity.runtime.newJavaMethod("Create an Arity object from the given messages. The list of unevaluated messages given to this method will be used as if they were the arguments part of a DefaultMethod definition.", new TypeCheckingJavaMethod("from") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(arity)
                    .withRestUnevaluated("arguments")
                    .getArguments();
                
                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject self, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = message.getArguments();
                    if (args.size() == 0) { 
                        return takingNothing(self);
                    }
                    DefaultArgumentsDefinition def = DefaultArgumentsDefinition.createFrom(args, 0, args.size(), message, on, context);
                    return getArity(self, def);
                }
            }));

        arity.registerMethod(arity.runtime.newJavaMethod("returns the names for positional arguments", new TypeCheckingJavaMethod("positionals") {
                
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .withOptionalPositional("includeOptionals", "true")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Arity a = (Arity) IokeObject.data(on);
                    List<Object> names = new ArrayList<Object>();
                    boolean includeOptional = args.isEmpty() ? true : IokeObject.isTrue(args.get(0));
                    if (a.argumentsDefinition != null) {
                        for(DefaultArgumentsDefinition.Argument argument : a.argumentsDefinition.getArguments()) {
                            if(argument instanceof DefaultArgumentsDefinition.KeywordArgument) { continue; }
                            if (argument instanceof DefaultArgumentsDefinition.OptionalArgument) {
                                if (includeOptional) {
                                    names.add(method.runtime.getSymbol(argument.getName()));
                                }
                            } else {
                                names.add(method.runtime.getSymbol(argument.getName()));
                            }
                        }
                    }
                    return method.runtime.newList(names);
                }
            }));

        arity.registerMethod(arity.runtime.newJavaMethod("returns the names for keyword arguments", new TypeCheckingJavaMethod.WithNoArguments("keywords", arity) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Arity a = (Arity) IokeObject.data(on);
                    List<Object> names = new ArrayList<Object>();
                    if (a.argumentsDefinition != null) {
                        for(String name : a.argumentsDefinition.getKeywords()) {
                            names.add(method.runtime.getSymbol(name.substring(0, name.length() - 1)));
                        }
                    }
                    return method.runtime.newList(names);
                }
            }));

        arity.registerMethod(arity.runtime.newJavaMethod("returns the symbol name for the krest argument.", new TypeCheckingJavaMethod.WithNoArguments("krest", arity) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Arity a = (Arity) IokeObject.data(on);
                    if (a.argumentsDefinition != null) {
                        String name = a.argumentsDefinition.getKrestName();
                        if (name == null) { 
                            return method.runtime.nil;
                        } else {
                            return method.runtime.getSymbol(name);
                        }
                    } else {
                        return method.runtime.nil;
                    }
                }
            }));

        arity.registerMethod(arity.runtime.newJavaMethod("returns the symbol name for the rest argument.", new TypeCheckingJavaMethod.WithNoArguments("rest", arity) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Arity a = (Arity) IokeObject.data(on);
                    if (a.argumentsDefinition != null) {
                        String name = a.argumentsDefinition.getRestName();
                        if (name == null) {
                            return method.runtime.nil;
                        } else {
                            return method.runtime.getSymbol(name);
                        }
                    } else {
                        return method.runtime.nil;
                    }
                }
            }));

        arity.registerMethod(arity.runtime.newJavaMethod("returns the text representation of this arity", new TypeCheckingJavaMethod.WithNoArguments("asText", arity) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Arity a = (Arity) IokeObject.data(on);
                    if (a.taking == Taking.Everything) {
                        return method.runtime.newText("...");
                    } else if (a.taking == Taking.Nothing) {
                        return method.runtime.newText("");
                    }
                    return method.runtime.newText(a.argumentsDefinition.getCode(false));
                }
            }));

    }
    
}

File: src/ikc/main/Ioke.Lang/Arity.cs

namespace Ioke.Lang {
    using System.Collections;
    using Ioke.Lang.Util;

    public class Arity : IokeData {
        private enum Taking { Unset, Nothing, Everything }

        private static IokeObject GetArity(IokeObject self, Taking thing) {
            IokeObject obj = self.runtime.Arity.AllocateCopy(null, null);
            obj.MimicsWithoutCheck(self.runtime.Arity);
            obj.Data = new Arity(thing);
            return obj;
        }

        public static IokeObject GetArity(IokeObject self, DefaultArgumentsDefinition def) {
            if(def == null || def.IsEmpty) {
                return IokeObject.As(TakingNothing(self), self.runtime.Arity);
            }
            IokeObject obj = self.runtime.Arity.AllocateCopy(null, null);
            obj.MimicsWithoutCheck(self.runtime.Arity);
            obj.Data = new Arity(def);
            return obj;
        }

        private DefaultArgumentsDefinition argumentsDefinition;
        private Taking taking = Taking.Unset;

        private Arity(Taking taking) {
            this.taking = taking;
        }

        public Arity(DefaultArgumentsDefinition argumentsDefinition) {
            if(argumentsDefinition == null || argumentsDefinition.IsEmpty) {
                this.taking = Taking.Nothing;
            } else {
                this.argumentsDefinition = argumentsDefinition;
            }
        }

        public static object TakingNothing(IokeObject self) {
            return self.runtime.Arity.GetCell(null, null, "taking:nothing");
        }

        public static object TakingEverything(IokeObject self) {
            return self.runtime.Arity.GetCell(null, null, "taking:everything");
        }

        public override void Init(IokeObject obj) {
            obj.Kind = "Arity";
        
            obj.SetCell("taking:nothing", GetArity(obj, Taking.Nothing));
            obj.SetCell("taking:everything", GetArity(obj, Taking.Everything));
        
            obj.RegisterMethod(obj.runtime.NewNativeMethod("Create an Arity object from the given messages. The list of unevaluated messages given to this method will be used as if they were the arguments part of a DefaultMethod definition.", 
                                                           new TypeCheckingNativeMethod("from", TypeCheckingArgumentsDefinition.builder()
                                                                                        .ReceiverMustMimic(obj)
                                                                                        .WithRestUnevaluated("arguments")
                                                                                        .Arguments,
                                                                                        (self, on, args, keywords, context, message) => {
                                                                                            if (message.Arguments.Count == 0) { 
                                                                                                return TakingNothing(self);
                                                                                            }
                                                                                            DefaultArgumentsDefinition def = DefaultArgumentsDefinition.CreateFrom(message.Arguments, 0, message.Arguments.Count, message, on, context);
                                                                                            return GetArity(self, def);
                                                                                        })));

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns the names for positional arguments", 
                                                           new TypeCheckingNativeMethod("positionals", TypeCheckingArgumentsDefinition.builder()
                                                                                        .WithOptionalPositional("includeOptionals", "true")
                                                                                        .Arguments,
                                                                                        (method, on, args, keywords, context, message) => {
                                                                                            Arity a = (Arity)IokeObject.dataOf(on);
                                                                                            var names = new SaneArrayList();
                                                                                            bool includeOptional = args.Count == 0 ? true : IokeObject.IsObjectTrue(args[0]);
                                                                                            if(a.argumentsDefinition != null) {
                                                                                                foreach(DefaultArgumentsDefinition.Argument argument in a.argumentsDefinition.Arguments) {
                                                                                                    if(argument is DefaultArgumentsDefinition.KeywordArgument) { continue; }
                                                                                                    if(argument is DefaultArgumentsDefinition.OptionalArgument) {
                                                                                                        if(includeOptional) {
                                                                                                            names.Add(method.runtime.GetSymbol(argument.Name));
                                                                                                        }
                                                                                                    } else {
                                                                                                        names.Add(method.runtime.GetSymbol(argument.Name));
                                                                                                    }
                                                                                                }
                                                                                            }
                                                                                            return method.runtime.NewList(names);
                                                                                        })));

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns the names for keyword arguments", 
                                                           new TypeCheckingNativeMethod.WithNoArguments("keywords", obj,
                                                                                                        (method, on, args, keywords, context, message) => {
                                                                                                            Arity a = (Arity)IokeObject.dataOf(on);
                                                                                                            var names = new SaneArrayList();
                                                                                                            if(a.argumentsDefinition != null) {
                                                                                                                foreach(string name in a.argumentsDefinition.Keywords) {
                                                                                                                    names.Add(method.runtime.GetSymbol(name.Substring(0, name.Length - 1)));
                                                                                                                }
                                                                                                            }
                                                                                                            return method.runtime.NewList(names);
                                                                                                        })));

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns the symbol name for the krest argument.", 
                                                           new TypeCheckingNativeMethod.WithNoArguments("krest", obj, 
                                                                                                        (method, on, args, keywords, context, message) => {
                                                                                                            Arity a = (Arity)IokeObject.dataOf(on);
                                                                                                            if(a.argumentsDefinition != null) {
                                                                                                                string name = a.argumentsDefinition.KrestName;
                                                                                                                if(name == null) { 
                                                                                                                    return method.runtime.nil;
                                                                                                                } else {
                                                                                                                    return method.runtime.GetSymbol(name);
                                                                                                                }
                                                                                                            } else {
                                                                                                                return method.runtime.nil;
                                                                                                            }
                                                                                                        })));

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns the symbol name for the rest argument.", 
                                                           new TypeCheckingNativeMethod.WithNoArguments("rest", obj,
                                                                                                        (method, on, args, keywords, context, message) => {
                                                                                                            Arity a = (Arity)IokeObject.dataOf(on);
                                                                                                            if(a.argumentsDefinition != null) {
                                                                                                                string name = a.argumentsDefinition.RestName;
                                                                                                                if(name == null) {
                                                                                                                    return method.runtime.nil;
                                                                                                                } else {
                                                                                                                    return method.runtime.GetSymbol(name);
                                                                                                                }
                                                                                                            } else {
                                                                                                                return method.runtime.nil;
                                                                                                            }
                                                                                                        })));

            obj.RegisterMethod(obj.runtime.NewNativeMethod("returns the text representation of this arity", 
                                                           new TypeCheckingNativeMethod.WithNoArguments("asText", obj,
                                                                                                        (method, on, args, keywords, context, message) => {
                                                                                                            Arity a = (Arity) IokeObject.dataOf(on);
                                                                                                            if (a.taking == Taking.Everything) {
                                                                                                                return method.runtime.NewText("...");
                                                                                                            } else if (a.taking == Taking.Nothing) {
                                                                                                                return method.runtime.NewText("");
                                                                                                            }
                                                                                                            return method.runtime.NewText(a.argumentsDefinition.GetCode(false));
                                                                                                        })));
        }
    }
}
